File: crates/note-daemon/src/rpc.rs
use std::fmt;

use chrono::NaiveDate;
use note_core::{DateRange, TaskStatus};
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

pub(crate) type RpcResult<T> = Result<T, RpcError>;

pub(crate) fn to_json<T: Serialize>(value: T) -> RpcResult<Value> {
    serde_json::to_value(value).map_err(|err| RpcError::internal(err.to_string()))
}

pub(crate) fn parse_params<T: DeserializeOwned>(params: Option<Value>) -> RpcResult<T> {
    let value = params.unwrap_or_else(|| json!({}));
    serde_json::from_value(value).map_err(|err| RpcError::invalid_params(err.to_string()))
}

pub(crate) fn parse_range(start: String, end: String) -> RpcResult<DateRange> {
    let start = parse_date(&start)?;
    let end = parse_date(&end)?;
    Ok(DateRange { start, end })
}

pub(crate) fn parse_date(input: &str) -> RpcResult<NaiveDate> {
    NaiveDate::parse_from_str(input, "%Y-%m-%d")
        .map_err(|_| RpcError::invalid_params(format!("invalid date '{}'", input)))
}

#[derive(Deserialize)]
pub(crate) struct RpcRequest {
    pub jsonrpc: Option<String>,
    pub id: Option<Value>,
    pub method: String,
    pub params: Option<Value>,
}

#[derive(Serialize)]
pub(crate) struct RpcResponse {
    jsonrpc: &'static str,
    id: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    error: Option<RpcErrorBody>,
}

#[derive(Serialize)]
pub(crate) struct RpcErrorBody {
    code: i32,
    message: String,
}

impl RpcResponse {
    pub(crate) fn result(id: Value, result: Value) -> Self {
        RpcResponse {
            jsonrpc: "2.0",
            id,
            result: Some(result),
            error: None,
        }
    }

    pub(crate) fn error(id: Value, error: RpcError) -> Self {
        RpcResponse {
            jsonrpc: "2.0",
            id,
            result: None,
            error: Some(RpcErrorBody {
                code: error.code.as_i32(),
                message: error.message,
            }),
        }
    }
}

#[derive(Clone, Copy, Debug)]
enum RpcErrorCode {
    ParseError = -32700,
    InvalidRequest = -32600,
    MethodNotFound = -32601,
    InvalidParams = -32602,
    InternalError = -32603,
    ServerError = -32000,
}

impl RpcErrorCode {
    fn as_i32(self) -> i32 {
        self as i32
    }
}

#[derive(Debug)]
pub(crate) struct RpcError {
    code: RpcErrorCode,
    message: String,
}

impl RpcError {
    fn new(code: RpcErrorCode, message: impl Into<String>) -> Self {
        RpcError {
            code,
            message: message.into(),
        }
    }

    pub(crate) fn parse_error(message: impl Into<String>) -> Self {
        RpcError::new(RpcErrorCode::ParseError, message)
    }

    pub(crate) fn invalid_request(message: impl Into<String>) -> Self {
        RpcError::new(RpcErrorCode::InvalidRequest, message)
    }

    pub(crate) fn invalid_params(message: impl Into<String>) -> Self {
        RpcError::new(RpcErrorCode::InvalidParams, message)
    }

    pub(crate) fn method_not_found(method: &str) -> Self {
        RpcError::new(
            RpcErrorCode::MethodNotFound,
            format!("unknown method '{}'", method),
        )
    }

    pub(crate) fn internal(message: impl Into<String>) -> Self {
        RpcError::new(RpcErrorCode::InternalError, message)
    }
}

impl fmt::Display for RpcError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} (code {})", self.message, self.code.as_i32())
    }
}

impl From<anyhow::Error> for RpcError {
    fn from(err: anyhow::Error) -> Self {
        RpcError::new(RpcErrorCode::ServerError, err.to_string())
    }
}

#[derive(Default, Deserialize)]
#[serde(default)]
pub(crate) struct ListTasksParams {
    pub status: Option<TaskStatus>,
    pub tags: Option<Vec<String>>,
    pub text_search: Option<String>,
    pub touched_since: Option<String>,
}

#[derive(Deserialize)]
pub(crate) struct TaskDetailParams {
    pub task_id: String,
}

#[derive(Deserialize)]
pub(crate) struct TagParams {
    pub tag: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct RangeParams {
    pub start: String,
    pub end: String,
}

#[derive(Deserialize)]
pub(crate) struct OpenDailyParams {
    pub date: String,
}

#[derive(Deserialize)]
pub(crate) struct NoteParams {
    pub note_id: String,
}

#[derive(Deserialize)]
pub(crate) struct WriteNoteParams {
    pub note_id: String,
    pub content: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parse_params_reports_invalid_input() {
        let err = parse_params::<RangeParams>(Some(json!({ "start": 123 }))).unwrap_err();
        assert_eq!(err.code.as_i32(), RpcErrorCode::InvalidParams.as_i32());
    }
}

File: internal/rpc/rpc.go
package rpc

import (
	"encoding/json"
	"fmt"

	"github.com/jack/yapper/go-note/internal/core"
)

// Request models the JSON-RPC 2.0 request envelope.
type Request struct {
	JSONRPC string           `json:"jsonrpc"`
	ID      *json.RawMessage `json:"id"`
	Method  string           `json:"method"`
	Params  json.RawMessage  `json:"params"`
}

// Response represents either a successful or failed JSON-RPC call.
type Response struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Result  interface{}     `json:"result,omitempty"`
	Error   *ErrorBody      `json:"error,omitempty"`
}

// ErrorBody matches the JSON-RPC error object.
type ErrorBody struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// ErrorCode enumerates JSON-RPC error codes.
type ErrorCode int

const (
	CodeParseError     ErrorCode = -32700
	CodeInvalidRequest ErrorCode = -32600
	CodeMethodNotFound ErrorCode = -32601
	CodeInvalidParams  ErrorCode = -32602
	CodeInternalError  ErrorCode = -32603
	CodeServerError    ErrorCode = -32000
)

// Error is the application-level error propagated to the RPC layer.
type Error struct {
	Code    ErrorCode
	Message string
}

func (e Error) Error() string {
	return fmt.Sprintf("%s (code %d)", e.Message, e.Code)
}

// ResponseResult builds a success response.
func ResponseResult(id json.RawMessage, result interface{}) Response {
	return Response{
		JSONRPC: "2.0",
		ID:      id,
		Result:  result,
	}
}

// ResponseError builds an error response.
func ResponseError(id json.RawMessage, err Error) Response {
	return Response{
		JSONRPC: "2.0",
		ID:      id,
		Error: &ErrorBody{
			Code:    int(err.Code),
			Message: err.Message,
		},
	}
}

// NullID returns a JSON null identifier placeholder.
func NullID() json.RawMessage {
	return json.RawMessage("null")
}

// ParseParams decodes params into the supplied struct, defaulting to {}.
func ParseParams[T any](params json.RawMessage) (T, Error) {
	var out T
	payload := params
	if len(payload) == 0 || string(payload) == "null" {
		payload = []byte("{}")
	}
	if err := json.Unmarshal(payload, &out); err != nil {
		return out, Error{Code: CodeInvalidParams, Message: err.Error()}
	}
	return out, Error{}
}

// ParseDate parses YYYY-MM-DD inputs into a core.Date.
func ParseDate(value string) (core.Date, Error) {
	date, err := core.ParseDate(value)
	if err != nil {
		return core.Date{}, Error{
			Code:    CodeInvalidParams,
			Message: fmt.Sprintf("invalid date %q", value),
		}
	}
	return date, Error{}
}

// ParseRange builds a DateRange from ISO date strings.
func ParseRange(start, end string) (core.DateRange, Error) {
	startDate, err := core.ParseDate(start)
	if err != nil {
		return core.DateRange{}, Error{
			Code:    CodeInvalidParams,
			Message: fmt.Sprintf("invalid start date %q", start),
		}
	}
	endDate, err := core.ParseDate(end)
	if err != nil {
		return core.DateRange{}, Error{
			Code:    CodeInvalidParams,
			Message: fmt.Sprintf("invalid end date %q", end),
		}
	}
	return core.DateRange{Start: startDate, End: endDate}, Error{}
}

// Convenience constructors for common error responses.
func ParseError(message string) Error {
	return Error{Code: CodeParseError, Message: message}
}

func InvalidRequest(message string) Error {
	return Error{Code: CodeInvalidRequest, Message: message}
}

func InvalidParams(message string) Error {
	return Error{Code: CodeInvalidParams, Message: message}
}

func MethodNotFound(method string) Error {
	return Error{Code: CodeMethodNotFound, Message: fmt.Sprintf("unknown method %q", method)}
}

func InternalError(message string) Error {
	return Error{Code: CodeInternalError, Message: message}
}

func ServerError(message string) Error {
	return Error{Code: CodeServerError, Message: message}
}

// Parameter payloads ---------------------------------------------------------

type ListTasksParams struct {
	Status       *core.TaskStatus `json:"status"`
	Tags         []string         `json:"tags"`
	TextSearch   *string          `json:"text_search"`
	TouchedSince *string          `json:"touched_since"`
}

type TaskDetailParams struct {
	TaskID string `json:"task_id"`
}

type TagParams struct {
	Tag string `json:"tag"`
}

type RangeParams struct {
	Start string `json:"start"`
	End   string `json:"end"`
}

type OpenDailyParams struct {
	Date string `json:"date"`
}

type NoteParams struct {
	NoteID string `json:"note_id"`
}

type WriteNoteParams struct {
	NoteID  string `json:"note_id"`
	Content string `json:"content"`
}
